ui/plugin-system/src/context/TimeRangeStateProvider.tsx

```diff
@@ -0,0 +1,59 @@
+// Copyright 2022 The Perses Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import React, { useState, useMemo, useCallback } from 'react';
+import { TimeRangeValue, AbsoluteTimeRange, toAbsoluteTimeRange, isRelativeTimeRange } from '@perses-dev/core';
+import { TimeRange, TimeRangeContext } from '../runtime/time-range';
+
+export interface TimeRangeProviderProps {
+  initialValue: TimeRangeValue;
+  children?: React.ReactNode;
+  onTimeRangeChange?: (e: TimeRangeValue) => void;
+}
+
+/**
+ * Provider implementation that supplies the TimeRangeState at runtime.
+ */
+export function TimeRangeStateProvider(props: TimeRangeProviderProps) {
+  const { initialValue, children, onTimeRangeChange } = props;
+
+  const [timeRange, setActiveTimeRange] = useState<TimeRangeValue>(initialValue);
```

Comment:
Back to my earlier comment, the whole state flow here still feels very confusing to me. As I'm looking (and I could be wrong, so apologies in advance), it looks like we have:

1. A `useState` out in the view with a time range (maybe only used as an initial time range, but not sure?)
2. Two `useState` calls in this provider
3. The values in the query string from react router out in the view

To me, it feels like "what is the currently selected time range" is all just one piece of state? There might be other versions we can derive from that (e.g. always converting the currently selected time range to absolute if it's relative), but still feels like there's only one piece of state. I just wonder if there's a way to simplify that flow so we really only have one "source of truth" here.

By "source of truth" I basically mean, only one `useState` or one `useQueryParams` or whatever it is that is responsible for getting/setting the value. That being said, I could also be totally misunderstanding what's going on or what the various pieces of state are for. 😂

Reply:
+1 to this. I think managing one piece is state is cleaner then deriving other state via useMemo

Reply:
Changed this back so now there isn't timeRange and absoluteTimeRange, I think a lot of the confusion is due to what I [described here](https://github.com/perses/perses/pull/530/#discussion_r972118025) about not wanting to make React Router v6 a required dependency in @perses-dev/dashboards, @perses-dev/plugin-system, etc. Ideally we would just be able to set the query params directly from TimeRangeControls / other components. 

Dang, maybe I should just go ahead and change to the QueryStringProvider approach we talked about before 🤔

Reply:
PR is now updated with the QueryStringProvider approach--useQueryParams is the single source of truth unless an app does not provide setQueryParams, then setTimeRange is used as a fallback.